feat: make Layout setting cumulative

Blankeos opened the pull request:
> [!WARNING]
> 
> This is just an attempt at a solution, I'm having blockers. Apparently, it's kind of tricky accomplish this in SolidJS: recursively wrapping a component with a dynamic array of components (layouts).
> 
> It renders correctly, but getting hydration errors.

Feel free to look into this and maybe attempt other solutions.

I'm using these as reference: 
- `MultiProvider` - Wrapping context recursively: https://github.com/solidjs-community/solid-primitives/blob/8e6089531dea15947d9c4671d764fe861413f2e5/packages/context/src/index.ts#L102
- Wrapping layout recursively: https://discord.com/channels/722131463138705510/1250852970589978654/1250881008375566357

<img width="699" alt="image" src="https://github.com/vikejs/vike-solid/assets/38070918/a0887702-1511-43d5-9ad0-c1604218f491">

brillout commented:
Would this work? 

```jsx
<Dynamic component={Layout1}>
   <Dynamic component={Layout2}>
     {props.children}
   </Dynamic>
</Dynamic>
```

Wouldn't the following do the trick then?

```jsx
let el = props.children
Layouts.forEach(Layout => {
  const child = el
  el = <Dynamic component={Layout}>{child}</Dynamic>
})
```

Blankeos commented:
@brillout yes that's what I'm trying to accomplish, that solution unfortunately doesn't work. Just tried it, it will cause hydration errors. For instance:

```tsx
function Layout(props: ParentProps) {
  const pageContext = usePageContext();
  
  let el = props.children;
  pageContext.config.Layout?.forEach((Layout: FlowComponent) => {
    const child = el;
    el = <Dynamic component={Layout}>{child}</Dynamic>;
  })

  return el;
}
```

That would work in React, but not in SolidJS. That's technically also the reason why it's tricky because Solid usually doesn't like "[early returns](https://stackoverflow.com/questions/78037797/why-cant-i-have-early-returns-in-solidjs#:~:text=Early%20returns%20do%20not%20work,Switch%20%2F%20Match%20for%20conditional%20rendering.)". Anything in Solid's function body before a `return` will only run once--Unless it's memoized or is a signal (`createMemo`, `createStore`, `createSignal`). A single component in SolidJS essentially only returns once. Doing stuff like this will render correctly **initially**, but lose out on reactivity. But if `getPageElement` only actually runs on the server, not on the client-side. Then this "maintaining reactivity" argument isn't really important because both just spit out the same HTML initially.

I tried a couple other solutions (this current PR using recursion), but no luck. This actually does maintain reactivity I think ([based on the other SolidJS libraries](https://github.com/solidjs-community/solid-primitives/blob/8e6089531dea15947d9c4671d764fe861413f2e5/packages/context/src/index.ts#L102) using that pattern, I linked them above).

### But here's an approach that didn't cause hydration errors though:
```tsx
function Layout(props: ParentProps) {
  const pageContext = usePageContext();

  return (
    <For each={pageContext.config.Layout?.reverse() ?? []}>
      {(Layout: FlowComponent) => (
        <Dynamic component={Layout}>{props.children}</Dynamic>
      )}
    </For>
  );
}
```

Weirdly, the client-side version likes this even if the component tree is ultimately looking like:
```tsx
<Layout1>
  {props.children}
</Layout2>
<Layout2>
  {props.children}
</Layout2>
```

### So, it actually got me thinking... Is `getPageElement` only running on the server? 

To confirm this, on the client-side app, I made the +config to `ssr: false`, then removed the `<Layout>` usage on `getPageElement` in vike-solid. The client-side does still actually try to render the layouts in the same weird way as above (NOTE: even when I removed the Layout on `getPageElement` already). Maybe there's something wrong with how the client-side renders it and causes these hydration errors? 

| Notice there's no hydration errors, there are layouts | Even though I removed the <Layout> on the getPageElement |
|--------|--------|
| <img width="955" alt="image" src="https://github.com/vikejs/vike-solid/assets/38070918/d15994e5-5468-43c7-9c8f-e0e69462d9b4"> | <img width="511" alt="image" src="https://github.com/vikejs/vike-solid/assets/38070918/eef3dc8c-4181-461d-95b5-ec4352ee9fca"> | 

**Can you point me into the right direction for investigating this behavior?**

brillout commented:
> Solid usually doesn't like "early returns" [...] will lose out on reactivity

I see. It's beyond my Solid knowledge but maybe Joël has some insights.

Blankeos commented:
Do you know what code controls the client-side rendering under the hood (Like when the config is set to: `ssr: false`)? I'm thinking it might be what's causing the hydration errors.

I don't think it's `getPageElement` since when I make changes on it, it doesn't get reflected on the rendered html on the frontend. Is it still controlled by `vike-solid` or just by `vike`?

I noticed there's something inherently weird with vike-solid's client-side rendering at the moment with cumulative layouts, I just can't figure out which part of the code is causing it. I did the same example on vike-react as well, it works as expected. Problem seems to only happen for vike-solid.

https://github.com/vikejs/vike-solid/assets/38070918/aa942a79-f6cf-4347-a0a3-4cf4cda76827

Blankeos commented:
### I've finally done it!!!! AHHHH! 
I don't know how I fixed it. Apparently it was `getPageElement` all along. 

The blockers I was facing were due to some:
- weird interoperation between pnpm and bun between my vike-solid and my local project when I `pnpm build` and `bun link`. (resolution was just to use pnpm)
- browser caching or something (not completely sure, but I just set Disable Cache on the browser during development)

Apparently the recursive solution worked!

https://github.com/vikejs/vike-solid/assets/38070918/fd049755-de72-4e54-b783-9795f2eed6c1

Should be ready for review now. Feel free to test it out here:
- Vike Frontend: https://github.com/Blankeos/solid-vike-nested-layouts

Blankeos commented:
Oh right, forgot. I might also have to add a "nested layouts" tab in the /examples folder similar to what you did on vike-react.

brillout commented:
Nice! :100:

Very much looking forward to have Nested Layouts also for `vike-solid` :sparkles: 

Since I ain't that familiar with Solid it's probably we wait on Joël's review next week.

Blankeos commented:
Just amended a change. Trimmed some unnecessary lines of code from [where I referenced it](https://github.com/solidjs-community/solid-primitives/blob/8e6089531dea15947d9c4671d764fe861413f2e5/packages/context/src/index.ts#L102).

Noticed something weird when it's `pnpm build` and `pnpm preview` though. Don't know if it's a vike bug (based on the logs). But I tried to reproduce it on react, and it doesn't happen for vike-react.

Wondering what could cause this though since it works perfectly on dev. But not on build.

I made a repro here: https://github.com/blankeos/solid-vike-nested-layouts
- Go to /dashboard (Nested layout 2-levels deep)
- Then go (Nested layout 1-level deep)

```
Error: [vike@0.4.176][Bug] You stumbled upon a Vike bug. Go to https://github.com/vikejs/vike/issues/new and copy-paste this error. A maintainer will fix the bug (usually under 24 hours).
    at de (entry-client-routing.ByC2kXXH.js:2:1671)
    at s (entry-client-routing.ByC2kXXH.js:3:841)
    at entry-client-routing.ByC2kXXH.js:3:19937
    at Array.map (<anonymous>)
    at entry-client-routing.ByC2kXXH.js:3:19898
    at Array.forEach (<anonymous>)
    at Tr (entry-client-routing.ByC2kXXH.js:3:19488)
    at $e (entry-client-routing.ByC2kXXH.js:3:49912)
    at async p (entry-client-routing.ByC2kXXH.js:4:12564)
    at async Q (entry-client-routing.ByC2kXXH.js:4:11817)
```

<img width="794" alt="image" src="https://github.com/vikejs/vike-solid/assets/38070918/458aad9e-f5f5-4f27-a332-51db94ba2289">

brillout commented:
> Error: [vike@0.4.176][Bug] You stumbled upon a Vike bug.

I cannot reproduce with following steps:
1. `pnpm run dev` 
2. Click `Dashboard`
3. Click `Settings`

I don't see any error in the browser dev console nor in the terminal.

Blankeos commented:
> I cannot reproduce with following steps:
> 
> `pnpm dev`

On dev it works fine. It's when it's built that's a problem.

Have you tried `pnpm build` `pnpm preview`?

brillout commented:
Indeed I can reproduce now. It seems to be a Vike bug, let me dig & fix.

Blankeos commented:
> Instead of directly manipulating `pageContext.config.Layout`, can you manipulating a copy instead?

Interesting. Thanks for investigating! Although, `.toReversed()` is an immutable operation and doesn't really manipulate the original array. Just returns a new one, so it's a bit weird.

I already create a copy called layoutStore.

I'll see what I can do about it again later. Thanks for replying on Discord btw.

brillout commented:
If you inject a `console.log()` before the `assert()` fails (third line in the stack trace), you'll see that there is some Solid proxy going on.

Blankeos commented:
I'm kinda stuck on this guys. 

Right now, it works perfectly on `pnpm dev` but not in `pnpm build && pnpm preview`. Can't see errors anymore. Currently developing on the `basic` example.

brillout commented:
Ryan Carniato had a quick look at it: https://x.com/RyanCarniato/status/1802764667754013109.

Blankeos commented:
Great news @brillout , I think I've finally done it! I've tested mostly always on `pnpm build` and `pnpm preview`. It's finally working the way it's expected. Do test it if you have time.

Summary of how I got to the resolution:
- Upgrading the solid-js version from `1.8.16` to `1.8.17` inside the `example/basic` seemed to fix the issue where the page doesn't render when navigating. (I think this is just a version mismatch during pnpm link so it gets confused that two solid-js versions are installed in the project)

- It seemed that the `reconcile` inside `onRenderClient` was causing the problems. So I removed it and just do `setPageContext(pageContext)` directly. (Although, I'm not sure if there are any negative implications to this).
- I think doing `reconcile` on  an `[() => {}, () => {}]` data structure is what might be causing those problems because it doesn't know how to diff it probably.
- This time I just reconcile the `pageContext.config.Layout` specifically inside the `<Layout />` component so it can persist the state when it tries to re-evaluate the new layouts.
- I also reverse the Layouts in `onRenderClient` so it's ready to use in the `getPageElement`.

brillout commented:
Neat! :100: Let's see what Joël thinks.

(FYI Solid is quite [responsive to bugs](https://github.com/solidjs/solid/issues?q=is%3Aissue+is%3Aopen+label%3Abug), so I wouldn't hesitate reporting bugs.)

magne4000 commented:
@Blankeos I was testing a few things on my side, and also came to the conclusion that `reconcile` doesn't work as intended with `vike` context. I stumbled upon [an issue](https://github.com/solidjs/solid/issues/1867) that seems to explain why.

I also managed to kinda fix the issue on my side also by trying to clone `pageContext` (which is not that easy when you have functions and Proxies). Your solution is simpler, so we'll keep it, but if a similar issue arise, I think that cloning the whole `pageContext` would be a more robust solution.

magne4000 commented:
I [removed the usage of `toReversed()`](https://github.com/vikejs/vike-solid/pull/86/commits/697a9271241939a407df0f092f35f5fc87031638) in order to keep compatibility with node@18. We just [loop in reverse order instead](https://github.com/vikejs/vike-solid/blob/697a9271241939a407df0f092f35f5fc87031638/vike-solid/renderer/getPageElement.tsx#L29).

magne4000 commented:
With those last changes, I think we're good to merge! @Blankeos good for you too?

Blankeos commented:
Hi @magne4000! I see, I remember using `.toReversed()` initially because Solid stores prevent you from mutating the original array.

Were there no problems with using `.reverse()`? If not, I'm good to go as well!

magne4000 commented:
> Were there no problems with using `.reverse()`

I don't know I didn't try. But mutating `pageContext` should be avoided, so I would have used something like `[...array].reverse()` instead if needed.
